src/dune_rules/exe.ml

```diff
@@ -208,7 +208,7 @@ let build_and_link_many ~programs ~linkages ~promote ?link_args ?o_files
   let modules = Compilation_context.modules cctx in
   let dep_graphs = Dep_rules.rules cctx ~modules in
   Module_compilation.build_all cctx ~dep_graphs;
-  let link_time_code_gen = Link_time_code_gen.handle_special_libs cctx in
+  let link_time_code_gen = lazy (Link_time_code_gen.handle_special_libs cctx) in
```

Comment:
Just a though (for another discussion): this use of lazy is correct regarding the memoization framework. However, we have to read carefully the code to convince ourselves of it. Using the memoized lazy feels a bit overkill here.

I'm wondering if we couldn't introduce a version of `lazy` that is as lilght as the plain `lazy`, but with a runtime check to make sure we didn't cross a memoised function boundary. This would allow us to use `lazy` whenever we want without thinking too hard about it.